keps/0007-20180413-user-namespaces.md

```diff
@@ -0,0 +1,267 @@
+
+## Summary
+
+In order to run Pods with software which expects to run as root or
+with elevated privileges while still containing the processes and
+protecting both the Nodes and other Pods, Linux kernel mechanism of
+user namespaces can be used make the processes in the Pods view their
+environment as having the privileges, while on the host (Node) level
+these processes appear as without privileges, or with privileges only
+affecting processes in the same Pods.
+
+In this proposal we look at three options of supporting and exposing
+the feature, showing the way of gradual deployment of the feature
+based both on the capabilities of the underlying Container Runtime,
+and to give admins and users mechanism to enable the feature
+gradually, to allow for testing and discovering regressions in their
+payloads.
+
+## Motivation
+
+From user_namespaces(7):
+> User namespaces isolate security-related identifiers and attributes,
+> in particular, user IDs and group IDs, the root directory, keys,
+> and capabilities.  A process's user and group IDs can be different
+> inside and outside a user namespace.  In particular, a process can
+> have a normal unprivileged user ID outside a user namespace while
+> at the same time having a user ID of 0 inside the namespace;
+> in other words, the process has full privileges for operations
+> inside the user namespace, but is unprivileged for operations
+> outside the namespace.
+
+The purpose of using user namespaces in Kubernetes is to let the
+processes in Pods think they run as one uid set when in fact they run
+as different “real” uids on the Nodes.
+
+In this text, most everything said about uids can also be applied to gids.
+
+### Goals
+
+The goal is to be able to take container images that would all want to
+run as the same uid (0 for root, 48 for apache, some 997 for nginx)
+or set of uids (systemd-based multi-daemon container which starts as
+uid 0 and forks a number of service running as different uids) and run
+them without clashing with the same uids on the host, or among
+containers themselves. So the set of uids that the containers will need
+mapped is known before the container is started.
+
+The effort will be successful when it will be possible to use any
+container image, built using any set of uids, and run it in Kubernetes
+cluster safely with runtime uids not matching those in the image.
+Specifically, Pods with containers that need to be run as root (uid 0)
+should be runnable with non-root uids, and systemd-based containers
+need to be supported.
+
+### Non-Goals
+
+The goal is not to support use cases when the uids for processes get
+allocated in runtime and externally (IPA, Active Directory) for generic
+users, without specifying (hopefully narrow) range of uids to be mapped
+upfront, before the container gets started.
+
+## Proposal
+
+This is where we get down to the nitty gritty of what the proposal actually is.
+
+### User Stories / Options
```

Comment:
i think it may help describing this section in terms of distinct feature gates.

1.  GlobalUserNamespace
1.  NamedUserNamespace

For item 1, please expand upon the mechanism by which across CRI implementations the kubelet could discover the mapping.  This is similar to discussion from @filbranden on the idea of a "Runtime or Distro ships static configuration file" in `/etc/cri-runtime/xyz` so the kubelet can introspect in a common manner.

For item 2 and item 3, please expand on the API changes and CRI change required in this KEP.

Item 1 can deliver a lot of value with minimal end-user pain so I prefer to see that as the first phase.

Reply:
also just capturing here for others, we are aware of a number of problems that need to be resolved at much lower levels to support "run containers with namespace specific userns remapping" especially around management of container filesystem.  as a result, the kep should describe the near term of what is possible today, and what the long term would be if those issues were resolved.